A round of edits (#1)
ceastlund/state-of-ppxlib-presentation: proposal/proposal.tex

source code meta-programming (Camlp4/5). Since then, meta-programming is mostly
done on the parsetree and has incrementally evolved into a solid ecosystem,
enriching the OCaml developer experience. Two highlights of that evolvement:

> introducing parsetree migrations to gain individual cross-version compatibility

Reviewer: What would you think about "cross-compiler-version" here and everywhere else? I'm not sure whether it's clear to everyone that we're talking about compatibility regarding the compiler when only saying "cross-version". Do you think it would be clear to everyone?
```suggestion
introducing parsetree migrations to gain individual cross-compiler-version compatibility
```

\rightarrow expanded\ AST\xrightarrow{\text{typing}} typedtree
\end{gather*}

To extend the nodes, PPXs need to manipulate the AST data type. Changes to the
AST data type would often break PPXes that depended on a previous form.
% ceastlund: I don't know of earlier workarounds. I don't think _every_ ppx broke on
% _every_ version, but the problem accelerated over time.

> To improve cross-version compatibility, $ocaml\mbox{-}migrate\mbox{-}parsetree$

Reviewer: ```suggestion
To improve cross-compiler-version compatibility, $ocaml\mbox{-}migrate\mbox{-}parsetree$
```
?

(OMP)\cite{omp} was invented. It converts the compiler AST version to the
version used by the PPX, then applies the PPX, and then converts back. However,
downward conversions are only well-defined if the AST doesn't contain new

> features/nodes. So full cross-version compatibility is only given if

Reviewer: ```suggestion
features/nodes. So full cross-compiler-version compatibility is only given if
```
?

\begin{align*}
\text{PPX\ AST\ version} >= \text{compiler\ AST\ version}
\end{align*}
That meant that the OCaml PPX world started to lag behind the compiler releases.

Initially, each PPX was a separate binary. The one early exception was $ppx \_
deriving$, which orchestrates multiple type-based derivers into a single
rewriter.

In 2018, $ppxlib$\cite{ppxlib} was created, in part to solve these problems. It
exposes a single version of the AST, which is kept up to date with the latest

> compiler and is backward-compatible with previous releases. This consolidates

Reviewer: What do you mean by "backward-compatible" here? The `ppxlib` AST isn't backward-compatible, is it?

Author: I'm trying to convey that ppxlib does the down-conversion and up-conversion automatically. I was probably too terse.

Reviewer: Do you think it becomes too convoluted if we write that out?

```suggestion
In 2018, $ppxlib$\cite{ppxlib} was created, in part to solve these problems. It
exposes a single version of the AST, which is kept up to date with the latest
compiler, and which all PPXs are defined against. The $ppxlib$\cite{ppxlib}
driver does the conversions to and from that AST. This consolidates
```

Btw, size might have become a problem indeed by now. In theory, we're supposed to hand in 2 pages, but (including references) are at almost 4 pages now.

Reviewer: We can keep it at "backward-compatible" if you think for others it's clear what's meant. Or we can totally remove that mention
```suggestion
In 2018, $ppxlib$\cite{ppxlib} was created, in part to solve these problems. It
exposes a single version of the AST, which is kept up to date with the latest
compiler. This consolidates
```

the PPX ecosystem and upgrades most PPXes ``for free''.

The primary remaining problem is the few PPXes that get broken each time the AST
changes. Our initial proposal, presented at the OCaml workshop
2019\cite{workshop2019}, was to create a single AST that could contain fragments
from multiple versions at once. This proved too complex and was subsequently
abandoned. We now approach \hyperref[stability]{stability} with a combination of
technical and social solutions.

\subsection{Typedtree based}

There have also been efforts to introduce typedtree based meta-programming,

> especially $MetaOCaml$\cite{metaocaml}. MetaOCaml and PPXes are different

Reviewer 2: ```suggestion
especially \texttt{MetaOCaml}\cite{metaocaml}. MetaOCaml and PPXes are different
```

Author: I've used fixed-width for things you'd actually write in OCaml source code, like `Stdlib`. I wouldn't write MetaOCaml that way, we've used math-italics for just names of things.

> There have also been efforts to introduce typedtree based meta-programming,
> especially $MetaOCaml$\cite{metaocaml}. MetaOCaml and PPXes are different
> notions of metaprogramming with different goals and expressivity. MetaOCaml is
> type-safe, staged metaprogramming that can be used to specialize values at
> compile-time or run-time. PPXes offer more general code generation than
> specializing values, but lack MetaOCaml's ability to evaluate the code and its
> integration with the type system.

Reviewer: Thanks, that's way better! Should we also mention somewhere that MetaOCaml is only available on a fork of OCaml?
```suggestion
There have also been efforts to introduce typedtree based meta-programming,
especially $MetaOCaml$\cite{metaocaml}. MetaOCaml and PPXes are different
notions of metaprogramming with different goals and expressivity. MetaOCaml is
type-safe, staged metaprogramming that can be used to specialize values at
compile-time or run-time. PPXes offer more general code generation than
specializing values, but lack MetaOCaml's ability to evaluate the code and its
integration with the type system. MetaOCaml requires non-merged changes
to the compiler and therefore only works on a fork of OCaml.
```

Point \ref{responsibility}. above touches on an important topic: PPXs generate
the final representation of the program and so have an impact on how the
compiler and the editor tool $merlin$ analyze the program. That's why we define
clear hygiene rules\cite{hygiene} and try to create a culture in which PPX
authors are aware of the possible consequences. Some of these hygiene rules are
enforced technically by $ppxlib$, and others are currently only guidelines that
PPX authors should follow. The PPX mechanism within the compiler itself does not
enforce any technical notion of hygiene as in some macro
systems\cite{scheme_hygiene}.

\subsubsection*{Error handling}

> Errors in OCaml are most often handled by raising exceptions. Naive error
> handling within the $ppxlib$ driver would abort transformation of an entire
> program the first time an exception was raised. This means tools like $merlin$
> learn nothing about the shape of the transformed program, and debugging the
> problem becomes only more difficult.
> 
> Our first approach to addressing this was to provide tools for expressing errors
> as AST nodes, and encouraging PPX authors to use these instead of exceptions.
> This practice was slow to creep into the ecosystem, so we made the behavior
> automatic by handling exceptions inside $ppxlib$. Now an exception raised by a
> context-free transformation is caught, and its error message is stored as an AST
> node so that the rest of the program can be transformed.

Reviewer: Arrg, thanks for noticing and correcting this! We should make clear that raising interrupts the rest of the rewriting, though, and hence it's still recommended to embed errors instead of raising. Maybe something along the lines of
```suggestion
Errors in OCaml are most often handled by raising exceptions. Naive error
handling within the $ppxlib$ driver would abort transformation of an entire
program the first time an exception was raised. This means tools like $merlin$
learn nothing about the shape of the transformed program, and debugging the
problem becomes only more difficult.

Our first approach to addressing this was to provide tools for expressing errors
as AST nodes, and encouraging PPX authors to use these instead of exceptions.
This practice was slow to creep into the ecosystem, so we made the behavior
automatic by handling exceptions inside $ppxlib$. Now an exception raised by a
context-free transformation is caught, and its error message is stored as an AST
node inside the last AST of the rewriting process. So a raising PPX still means that
the rewriting flow is interrupted, and the `ppxlib` driver returns a partly rewritten
AST. Therefore, we still encourage PPX authors to embed errors instead of raising.
```
? What do you think?

Author: What stage of the rewriting is interrupted? The closest enclosing context-free rewrite? Or the whole thing?

Reviewer: The stage of the latest full AST. So in the case of a raising context-free rewriter, the AST from before the context-free phase.

Author: Oh. Hmm. We may want to consider catching things at the level of individual extension nodes.

\subsubsection*{Location invariants}

The \emph{Location} of a AST node reveals which part of the source code the node
corresponds to. We've run into some problems with transformed code, wherein
tools like $merlin$ cannot figure out which piece of transformed or duplicated
code to analyze when the user queries the source program.
%Locations are very important for the compiler/$merlin$ to point to the right
%place when reporting errors/warnings.

To resolve this issue, we've instituted two location invariants for the results
of PPX transformations. These can be configured in the $ppxlib$ driver as
automatic checks, or left to PPX authors to enforce.
\begin{itemize}
\item The locations of child nodes are contained by the locations of their parents.
\item The locations of sibling nodes don't overlap.

> \end{itemize}

Reviewer 2: Maybe add a mention on ghost locations?

```suggestion
\end{itemize}

It is also possible to escape those invariants by using "ghost locations". Using ghost locations for nodes that borrow their location to another node is a good practice, to avoid confusing Merlin.
```

Author: How about the following:

`It is possible to escape these invariants using "ghost locations". Locations can be annotated as "ghost" to direct tools to ignore them. This is useful for generated or duplicated code snippets that share locations. Each location should have a single non-ghost node so that tools know what to associate with that location.`

Reviewer: I think the last sentence might come across a bit misleading. Isn't it more like "each position should be contained in the locations of a unique sequence of parent-child nodes"?
